### PR TITLE
XAuth support for Linux clients

### PR DESCRIPTION
**Pull request by kshcherban**
Added support for Aggressive Mode that is being used by `vpnc`.
Documented Linux client setup on Ubuntu example.

### CODE REVIEW

File: vpnsetup.sh [hunk 1]
```diff
@@ -245,7 +247,7 @@ conn shared
   dpddelay=30
   dpdtimeout=120
   dpdaction=clear
-  ike=3des-sha1,3des-sha2,aes-sha1,aes-sha1;modp1024,aes-sha2,aes-sha2;modp1024
+  ike=3des-sha1,3des-sha1;modp1024,3des-sha2,aes-sha1,aes-sha1;modp1024,aes-sha2,aes-sha2;modp1024
```
[the remark this answers is not in the document]
Comment (kshcherban) [reply vpnsetup.sh]: vpnc doesn't work without those cyphers, don't know if it's the bug of LibreSwan or vpnc.

File: vpnsetup.sh [hunk 2]
```diff
@@ -245,7 +247,7 @@ conn shared
...
   phase2alg=3des-sha1,3des-sha2,aes-sha1,aes-sha2
```
[the remark this answers is not in the document]
Comment (kshcherban) [reply vpnsetup.sh]: That should work both with android and vpnc (desktop linux), tested on my server.

[the remark this answers is not in the document]
Comment (kshcherban) [reply vpnsetup.sh]: Updated, please check.

**Comment by hwdsl2** [comment]
@kshcherban Thank you again for contributing! I've decided not to merge these changes at this time because:

1. The limit of 16384 bytes for AWS instance user data, which the script is approaching
2. Connecting using `vpnc` from Linux in aggressive mode is not a common use case
3. The proposed solution requires "pam" authentication (which requires adding a user to the system). I haven't figured out how to instead use `authby=file` for this.
4. The `VPN_GROUP` could be confusing for people using the IPsec/L2TP or IPsec/XAuth main mode.
5. The IPsec aggressive mode is vulnerable to offline dictionary attacks on the IPsec PSK [1].

However, if the changes work for you, you're welcome to adapt the scripts to your needs.

[1] https://security.stackexchange.com/questions/76444/what-are-the-practical-risks-of-using-ike-aggressive-mode-with-a-pre-shared-key

**Comment by letoams** [comment]
On Fri, 16 Nov 2018, Lin Song wrote:

>  5. The IPsec aggressive mode is vulnerable to offline dictionary attacks on the IPsec PSK [1].

> [1] https://security.stackexchange.com/questions/76444/what-are-the-practical-risks-of-using-ike-aggressive-mode-with-a-pre-shared-key

Technically, Main Mode and IKEv2 are also vulnerable to this, see:

https://datatracker.ietf.org/meeting/103/materials/slides-103-ipsecme-psks-will-always-be-weak-00

But I agree that to reduce the script, it is best to only support  IKEv2
and IKEv1 L2TP/IPsec

Paul

**Pull request closed without merging**